README.md

```diff
@@ -0,0 +1,434 @@
+
+
+the flow when you subscribe / call run on a cached value is relatively simple:
+!["cached flow"](docs/CachedValue.svg)
+
+
+
+when you call run / subscribe on a combination `query`/`inputs` that was never retrieved before (or whose last try to retrive it ended up with a `Failure`), avenger will try to run the `Fetch` function resulting in a more complex flow:
+!["cached flow"](docs/UncachedOrErrorValue.svg)
+
+## composing queries
+you can build bigger queries from smaller ones in two ways:
+
+- by composing them with [**`compose`**](#compose): when you need your queries to be sequentially run with the results of one feeding the other, you can use `compose`:
+
+```ts
+import { compose } from 'avenger/lib/Query';
+
+type UserPreferences = { color: string };
+
+// note that the two ends of the composed functions must have compatible types
+declare function getUser(userId: number): TaskEither<Error, User>;
+declare function getUserPreferences(
+  user: User
+): TaskEither<Error, UserPreferences>;
+
+const userQuery: ObservableQuery<number, Error, User> = queryShallow(
+  getUser,
+  refetch
+);
+
+const preferencesQuery: ObservableQuery<
+  User,
+  Error,
+  UserPreferences
+> = queryShallow(getUserPreferences, refetch);
+
+const composition: Composition<number, Error, UserPreferences> = compose(
+  userQuery,
+  preferencesQuery
+);
+```
+
+- by grouping them with [**`product`**](#product): when you don't need to run the queries sequentially but would like to conveniently group them and treat them as if they where one you can use `product`:
+
+```ts
+const grouped = product({ myQuery, myQuery2 });
+```
+
+# commands
+Since now we only described how fetching of data is performed, when you need to update / insert data remotely you can make use of [**`command`**](#command):
+
+```ts
+declare function updateUserPreferences({
+  color: string
+}): TaskEither<Error, void>;
+
+const updatePreferencesCommand = command(updateUserPreferences, {
+  preferencesQuery
+});
+```
+
+`command` accepts `Fetch` function that will be used to modify the remote data source and a second parameter, a `Record` of `query`es that will be invalidated once the command is succesfully run:
+
+```ts
+/* when you call the command you can specify the input value corresponding
+to the Cache key that sould be invalidated as a second parameter */
+updatePreferencesCommand({ color: 'acquamarine' }, { preferencesQuery: 1 });
+```
+
+# React
+
+Avenger also exports some utilities to use with `React`
+
+## declareQueries
+
+`declareQueries` in an `HOC` builder, it lets you define the queries that you want to inject into a component and then creates a simple `HOC` to wrap it:
+
+```ts
+import { declareQueries } from 'avenger/lib/react';
+import { userPreferences } from './queries';
+
+const queries = declareQueries({ userPreferences });
+
+class MyComponent extends React.PureComponent<Props, State> {
+  render() {
+    return this.props.queries.fold(
```

Comment:
Since now we never talked about folding on queries and query states (loading/error/ready)

Reply:
there is a brief mention when I say "(...) that conveniently lets you fold (...)" and a link to the type of `QueryResult` that mentions it is a bifunctor and monad. I thought that for an experienced reader that would be enough, at the same time experieced readers are probably the only ones interested in this.. I will expand a little on that